tests/core/test_jose/test_jwe.py

```diff
@@ -166,78 +169,656 @@ def test_ase_gcm_jwe_invalid_key(self):
+                protected = {'alg': alg, 'enc': enc}
+                data = jwe.serialize_compact(protected, b'hello', key)
+                rv = jwe.deserialize_compact(data, key)
+                self.assertEqual(rv['payload'], b'hello')
+
+    def test_ecdh_es_decryption_with_public_key_fails(self):
         jwe = JsonWebEncryption()
         protected = {'alg': 'ECDH-ES', 'enc': 'A128GCM'}
+
         key = {
             "kty": "EC",
             "crv": "P-256",
-            "x": "gI0GAILBdu7T53akrFmMyGcsF3n5dO7MmwNBHKW5SV0",
-            "y": "SLW_xSffzlPWrHEVI30DHM_4egVwt3NQqeUD7nMFpps",
+            "x": "weNJy2HscCSM6AEDTDg04biOvhFhyyWvOHQfeF_PxMQ",
+            "y": "e8lnCO-AlStT-NJVX-crhB7QRYhiix03illJOVAOyck"
         }
         data = jwe.serialize_compact(protected, b'hello', key)
-        self.assertRaises(ValueError, jwe.deserialize_compact, data, key)
+        self.assertRaises(
+            ValueError,
+            jwe.deserialize_compact,
+            data, key
+        )
+
+    def test_ecdh_es_encryption_fails_if_key_curve_is_inappropriate(self):
+        jwe = JsonWebEncryption()
+        protected = {'alg': 'ECDH-ES', 'enc': 'A128GCM'}
 
-        key = OKPKey.generate_key('Ed25519', is_private=True)
+        key = OKPKey.generate_key('Ed25519', is_private=False)
         self.assertRaises(
             ValueError,
             jwe.serialize_compact,
             protected, b'hello', key
         )
 
+    def test_ecdh_1pu_key_agreement_computation_appx_a(self):
+        # https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04#appendix-A
+        alice_static_key = {
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "WKn-ZIGevcwGIyyrzFoZNBdaq9_TsqzGl96oc0CWuis",
+            "y": "y77t-RvAHRKTsSGdIYUfweuOvwrvDD-Q3Hv5J0fSKbE",
+            "d": "Hndv7ZZjs_ke8o9zXYo3iq-Yr8SewI5vrqd0pAvEPqg"
+        }
+        bob_static_key = {
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "weNJy2HscCSM6AEDTDg04biOvhFhyyWvOHQfeF_PxMQ",
+            "y": "e8lnCO-AlStT-NJVX-crhB7QRYhiix03illJOVAOyck",
+            "d": "VEmDZpDXXK8p8N0Cndsxs924q6nS1RXFASRl6BfUqdw"
+        }
+        alice_ephemeral_key = {
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "gI0GAILBdu7T53akrFmMyGcsF3n5dO7MmwNBHKW5SV0",
+            "y": "SLW_xSffzlPWrHEVI30DHM_4egVwt3NQqeUD7nMFpps",
+            "d": "0_NxaRPUMQoAJt50Gz8YiTr8gRTwyEaCumd-MToTmIo"
+        }
+
+        headers = {
+            "alg": "ECDH-1PU",
+            "enc": "A256GCM",
+            "apu": "QWxpY2U",
+            "apv": "Qm9i",
+            "epk": {
+                "kty": "EC",
+                "crv": "P-256",
+                "x": "gI0GAILBdu7T53akrFmMyGcsF3n5dO7MmwNBHKW5SV0",
+                "y": "SLW_xSffzlPWrHEVI30DHM_4egVwt3NQqeUD7nMFpps"
+            }
+        }
+
+        alg = JsonWebEncryption.ALG_REGISTRY['ECDH-1PU']
+
+        alice_static_key = alg.prepare_key(alice_static_key)
```

Comment:
It would be great to check that Ze, Zs, Z, derived key and other parameters are equal to test vector values from Appendix A.